## Exercises: Filter and query data

In [3]:
import pandas as pd
autos = pd.read_json("../Data/autos_json.json")

In [4]:
# I explore the dataset on my own
pd.set_option('display.max_columns', None)
autos.head()

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
0        std  convertible  3.47        21                9.0         2548   
1        std  convertible  3.47        21                9.0         2548   
2        std    hatchback  2.68        19                9.0         2823   
3        std        sedan  3.19        24               10.0         2337   
4        std        sedan  3.19        18                8.0         2824   

  drive-wheels engine-location  engine-size engine-type fuel-system fuel-type  \
0          rwd           front          130        dohc        mpfi       gas   
1          rwd           front          130        dohc        mpfi       gas   
2          rwd           front          152        ohcv        mpfi       gas   
3          fwd           front          109         ohc        mpfi       gas   
4          4wd           front          136         ohc        mpfi       gas   

   height  highway-mpg  horsepower  length         make  normalized-losses  \
0    48.8           27       111.0   168.8  alfa-romero                NaN   
1    48.8           27       111.0   168.8  alfa-romero                NaN   
2    52.4           26       154.0   171.2  alfa-romero                NaN   
3    54.3           30       102.0   176.6         audi              164.0   
4    54.3           22       115.0   176.6         audi              164.0   

  num-of-cylinders num-of-doors  peak-rpm    price  stroke  symboling  \
0             four          two    5000.0  13495.0    2.68          3   
1             four          two    5000.0  16500.0    2.68          3   
2              six          two    5000.0  16500.0    3.47          1   
3             four         four    5500.0  13950.0    3.40          2   
4             five         four    5500.0  17450.0    3.40          2   

   wheel-base  width  
0        88.6   64.1  
1        88.6   64.1  
2        94.5   65.5  
3        99.8   66.2  
4        99.4   66.4

In [7]:
rows, cols = autos.shape
print(f'Antal rader: {rows}, antal kolumner: {cols}')

Antal rader: 205, antal kolumner: 26


In [9]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   aspiration         205 non-null    object 
 1   body-style         205 non-null    object 
 2   bore               201 non-null    float64
 3   city-mpg           205 non-null    int64  
 4   compression-ratio  205 non-null    float64
 5   curb-weight        205 non-null    int64  
 6   drive-wheels       205 non-null    object 
 7   engine-location    205 non-null    object 
 8   engine-size        205 non-null    int64  
 9   engine-type        205 non-null    object 
 10  fuel-system        205 non-null    object 
 11  fuel-type          205 non-null    object 
 12  height             205 non-null    float64
 13  highway-mpg        205 non-null    int64  
 14  horsepower         203 non-null    float64
 15  length             205 non-null    float64
 16  make               205 non

In [38]:
autos.columns

Index(['aspiration', 'body-style', 'bore', 'city-mpg', 'compression-ratio',
       'curb-weight', 'drive-wheels', 'engine-location', 'engine-size',
       'engine-type', 'fuel-system', 'fuel-type', 'height', 'highway-mpg',
       'horsepower', 'length', 'make', 'normalized-losses', 'num-of-cylinders',
       'num-of-doors', 'peak-rpm', 'price', 'stroke', 'symboling',
       'wheel-base', 'width'],
      dtype='object')

In [39]:
autos.index

RangeIndex(start=0, stop=205, step=1)

In [59]:
# kollar vilka värden som 'aspiration' har i data setet:
autos['aspiration'].unique()

array(['std', 'turbo'], dtype=object)

In [61]:
# Hur många cylindrar finns som max i datasetet?
autos['num-of-cylinders'].unique()


array(['four', 'six', 'five', 'three', 'twelve', 'two', 'eight'],
      dtype=object)

In [66]:
# Listar alla bilar med 'two' cylindrar
autos[autos['num-of-cylinders'] == 'two']

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
55        std  hatchback   NaN        17                9.4         2380   
56        std  hatchback   NaN        17                9.4         2380   
57        std  hatchback   NaN        17                9.4         2385   
58        std  hatchback   NaN        16                9.4         2500   

   drive-wheels engine-location  engine-size engine-type fuel-system  \
55          rwd           front           70       rotor        4bbl   
56          rwd           front           70       rotor        4bbl   
57          rwd           front           70       rotor        4bbl   
58          rwd           front           80       rotor        mpfi   

   fuel-type  height  highway-mpg  horsepower  length   make  \
55       gas    49.6           23       101.0   169.0  mazda   
56       gas    49.6           23       101.0   169.0  mazda   
57       gas    49.6           23       101.0   169.0  mazda   
58       gas    49.6           23       135.0   169.0  mazda   

    normalized-losses num-of-cylinders num-of-doors  peak-rpm    price  \
55              150.0              two          two    6000.0  10945.0   
56              150.0              two          two    6000.0  11845.0   
57              150.0              two          two    6000.0  13645.0   
58              150.0              two          two    6000.0  15645.0   

    stroke  symboling  wheel-base  width  
55     NaN          3        95.3   65.7  
56     NaN          3        95.3   65.7  
57     NaN          3        95.3   65.7  
58     NaN          3        95.3   65.7

In [70]:
# Kollar om 'ford' finns med i datasetet
autos['make'].unique() # returnerar en lista med alla märken

'ford' in autos['make'].unique()

False

In [88]:
# filtrerar fram alla volvo med query metoden
autos.query("make == 'volvo'")

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
194        std      sedan  3.78        23                9.5         2912   
195        std      wagon  3.78        23                9.5         3034   
196        std      sedan  3.78        24                9.5         2935   
197        std      wagon  3.78        24                9.5         3042   
198      turbo      sedan  3.62        17                7.5         3045   
199      turbo      wagon  3.62        17                7.5         3157   
200        std      sedan  3.78        23                9.5         2952   
201      turbo      sedan  3.78        19                8.7         3049   
202        std      sedan  3.58        18                8.8         3012   
203      turbo      sedan  3.01        26               23.0         3217   
204      turbo      sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type fuel-system  \
194          rwd           front          141         ohc        mpfi   
195          rwd           front          141         ohc        mpfi   
196          rwd           front          141         ohc        mpfi   
197          rwd           front          141         ohc        mpfi   
198          rwd           front          130         ohc        mpfi   
199          rwd           front          130         ohc        mpfi   
200          rwd           front          141         ohc        mpfi   
201          rwd           front          141         ohc        mpfi   
202          rwd           front          173        ohcv        mpfi   
203          rwd           front          145         ohc         idi   
204          rwd           front          141         ohc        mpfi   

    fuel-type  height  highway-mpg  horsepower  length   make  \
194       gas    56.2           28       114.0   188.8  volvo   
195       gas    57.5           28       114.0   188.8  volvo   
196       gas    56.2           28       114.0   188.8  volvo   
197       gas    57.5           28       114.0   188.8  volvo   
198       gas    56.2           22       162.0   188.8  volvo   
199       gas    57.5           22       162.0   188.8  volvo   
200       gas    55.5           28       114.0   188.8  volvo   
201       gas    55.5           25       160.0   188.8  volvo   
202       gas    55.5           23       134.0   188.8  volvo   
203    diesel    55.5           27       106.0   188.8  volvo   
204       gas    55.5           25       114.0   188.8  volvo   

     normalized-losses num-of-cylinders num-of-doors  peak-rpm    price  \
194              103.0             four         four    5400.0  12940.0   
195               74.0             four         four    5400.0  13415.0   
196              103.0             four         four    5400.0  15985.0   
197               74.0             four         four    5400.0  16515.0   
198              103.0             four         four    5100.0  18420.0   
199               74.0             four         four    5100.0  18950.0   
200               95.0             four         four    5400.0  16845.0   
201               95.0             four         four    5300.0  19045.0   
202               95.0              six         four    5500.0  21485.0   
203               95.0              six         four    4800.0  22470.0   
204               95.0             four         four    5400.0  22625.0   

     stroke  symboling  wheel-base  width  
194    3.15         -2       104.3   67.2  
195    3.15         -1       104.3   67.2  
196    3.15         -2       104.3   67.2  
197    3.15         -1       104.3   67.2  
198    3.15         -2       104.3   67.2  
199    3.15         -1       104.3   67.2  
200    3.15         -1       109.1   68.9  
201    3.15         -1       109.1   68.8  
202    2.87         -1       109.1   68.9  
203    3.40         -1       109.1   68.9  
204    3.15         -1       109.1  

In [16]:
# undersöker mer om pris på volvo
import math
average_volvo_price = autos[autos['make'] == 'volvo']['price'].mean()


18063.18181818182


In [91]:
# filtrerar fram alla volvo med boolean mask
autos.loc[autos['make'] == 'volvo',:]

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
194        std      sedan  3.78        23                9.5         2912   
195        std      wagon  3.78        23                9.5         3034   
196        std      sedan  3.78        24                9.5         2935   
197        std      wagon  3.78        24                9.5         3042   
198      turbo      sedan  3.62        17                7.5         3045   
199      turbo      wagon  3.62        17                7.5         3157   
200        std      sedan  3.78        23                9.5         2952   
201      turbo      sedan  3.78        19                8.7         3049   
202        std      sedan  3.58        18                8.8         3012   
203      turbo      sedan  3.01        26               23.0         3217   
204      turbo      sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type fuel-system  \
194          rwd           front          141         ohc        mpfi   
195          rwd           front          141         ohc        mpfi   
196          rwd           front          141         ohc        mpfi   
197          rwd           front          141         ohc        mpfi   
198          rwd           front          130         ohc        mpfi   
199          rwd           front          130         ohc        mpfi   
200          rwd           front          141         ohc        mpfi   
201          rwd           front          141         ohc        mpfi   
202          rwd           front          173        ohcv        mpfi   
203          rwd           front          145         ohc         idi   
204          rwd           front          141         ohc        mpfi   

    fuel-type  height  highway-mpg  horsepower  length   make  \
194       gas    56.2           28       114.0   188.8  volvo   
195       gas    57.5           28       114.0   188.8  volvo   
196       gas    56.2           28       114.0   188.8  volvo   
197       gas    57.5           28       114.0   188.8  volvo   
198       gas    56.2           22       162.0   188.8  volvo   
199       gas    57.5           22       162.0   188.8  volvo   
200       gas    55.5           28       114.0   188.8  volvo   
201       gas    55.5           25       160.0   188.8  volvo   
202       gas    55.5           23       134.0   188.8  volvo   
203    diesel    55.5           27       106.0   188.8  volvo   
204       gas    55.5           25       114.0   188.8  volvo   

     normalized-losses num-of-cylinders num-of-doors  peak-rpm    price  \
194              103.0             four         four    5400.0  12940.0   
195               74.0             four         four    5400.0  13415.0   
196              103.0             four         four    5400.0  15985.0   
197               74.0             four         four    5400.0  16515.0   
198              103.0             four         four    5100.0  18420.0   
199               74.0             four         four    5100.0  18950.0   
200               95.0             four         four    5400.0  16845.0   
201               95.0             four         four    5300.0  19045.0   
202               95.0              six         four    5500.0  21485.0   
203               95.0              six         four    4800.0  22470.0   
204               95.0             four         four    5400.0  22625.0   

     stroke  symboling  wheel-base  width  
194    3.15         -2       104.3   67.2  
195    3.15         -1       104.3   67.2  
196    3.15         -2       104.3   67.2  
197    3.15         -1       104.3   67.2  
198    3.15         -2       104.3   67.2  
199    3.15         -1       104.3   67.2  
200    3.15         -1       109.1   68.9  
201    3.15         -1       109.1   68.8  
202    2.87         -1       109.1   68.9  
203    3.40         -1       109.1   68.9  
204    3.15         -1       109.1  

In [96]:
# hitta den billigaste volvon:
df_volvo = autos[autos['make'] == 'volvo']
cheapest_volvo = df_volvo['price'].min()
print(f'Den billigaste volvon kostar {cheapest_volvo} usd.')

Den billigaste volvon kostar 12940.0 usd.


Leta upp den dyraste volvon med fyra cylindrar

In [21]:
autos[(autos['make'] == 'volvo') & (autos['num-of-cylinders'] == 'four')]['price'].max()

22625.0

In [27]:
autos.query("make == 'volvo' and `num-of-cylinders` == 'four'")

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
194        std      sedan  3.78        23                9.5         2912   
195        std      wagon  3.78        23                9.5         3034   
196        std      sedan  3.78        24                9.5         2935   
197        std      wagon  3.78        24                9.5         3042   
198      turbo      sedan  3.62        17                7.5         3045   
199      turbo      wagon  3.62        17                7.5         3157   
200        std      sedan  3.78        23                9.5         2952   
201      turbo      sedan  3.78        19                8.7         3049   
204      turbo      sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type  ...   make  \
194          rwd           front          141         ohc  ...  volvo   
195          rwd           front          141         ohc  ...  volvo   
196          rwd           front          141         ohc  ...  volvo   
197          rwd           front          141         ohc  ...  volvo   
198          rwd           front          130         ohc  ...  volvo   
199          rwd           front          130         ohc  ...  volvo   
200          rwd           front          141         ohc  ...  volvo   
201          rwd           front          141         ohc  ...  volvo   
204          rwd           front          141         ohc  ...  volvo   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
194             103.0              four          four    5400.0  12940.0   
195              74.0              four          four    5400.0  13415.0   
196             103.0              four          four    5400.0  15985.0   
197              74.0              four          four    5400.0  16515.0   
198             103.0              four          four    5100.0  18420.0   
199              74.0              four          four    5100.0  18950.0   
200              95.0              four          four    5400.0  16845.0   
201              95.0              four          four    5300.0  19045.0   
204              95.0              four          four    5400.0  22625.0   

    stroke  symboling wheel-base width  
194   3.15         -2      104.3  67.2  
195   3.15         -1      104.3  67.2  
196   3.15         -2      104.3  67.2  
197   3.15         -1      104.3  67.2  
198   3.15         -2      104.3  67.2  
199   3.15         -1      104.3  67.2  
200   3.15         -1      109.1  68.9  
201   3.15         -1      109.1  68.8  
204   3.15         -1      109.1  68.9  

[9 rows x 26 columns]

In [ ]:
# dyraste bilen i datasetet, returnerar hela raden
autos.loc[autos['price'] == autos['price'].max()]

**Use the "autos" dataset to answer the following questions:**

**Q1:** How many cars from Jaguar are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=3
&nbsp;&nbsp;&nbsp;price=36000.0
</details>

In [4]:
# Boolean mask för att ta reda på hur många bilar från Jaguar:
nbr_jaguars = len(autos[autos['make'] == 'jaguar'])
print(f'Number of Jaguars: {nbr_jaguars}')

Number of Jaguars: 3


In [5]:
# Query metoden för att ta reda på hur många Jaguar det finns:
nbr_jaguars2 = len(autos.query('make == "jaguar"'))
#nbr_jaguars2 = autos.query('make == "jaguar"').count()
print(f'Det finns {nbr_jaguars2} Jaguar i datasetet.')

Det finns 3 Jaguar i datasetet.


In [ ]:
# Dyraste Jaguaren, boolean mask

In [75]:
#Vad kostar den dyraste Jaguaren:
#Query metoden:
jaguars = autos.query("make == 'jaguar'")
jaguars['price'].max()

36000.0

**Q2:** How many cars from Toyota are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=32
&nbsp;&nbsp;&nbsp;price=17669.0
</details>

In [86]:
# Number of Toyotas
nbr_toyotas = autos['make'][autos['make'] == 'toyota'].count()
print(f'There are {nbr_toyotas} Toyotas.')

There are 32 Toyotas.


In [104]:
# What is the price of the most expensive Toyota?
df_toyotas = autos[autos['make'] == 'toyota']
most_expensive_toyota = df_toyotas['price'].max()
print(most_expensive_toyota)

17669.0


In [106]:
# Dyraste toyotan med en rad, boolean mask. Boolean mask returnerar hela raden för den aktuella bilen.
autos[(autos['make'] == 'toyota') & (autos['price'] == autos[autos['make'] == 'toyota']['price'].max())]

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
172        std  convertible  3.62        24                9.3         2975   

    drive-wheels engine-location  engine-size engine-type fuel-system  \
172          rwd           front          146         ohc        mpfi   

    fuel-type  height  highway-mpg  horsepower  length    make  \
172       gas    53.0           30       116.0   176.2  toyota   

     normalized-losses num-of-cylinders num-of-doors  peak-rpm    price  \
172              134.0             four          two    4800.0  17669.0   

     stroke  symboling  wheel-base  width  
172     3.5          2        98.4   65.6

In [108]:
# Dyraste Toyotan med en rad boolean mask:
autos[autos['make'] == 'toyota']['price'].max()

17669.0

**Q3:** What is the length, width and height of the most expensive car in the entire dataset?

<details>
<summary>Answer</summary>
<br>&nbsp;&nbsp;&nbsp;length=199.2
&nbsp;&nbsp;&nbsp;width=72.0
&nbsp;&nbsp;&nbsp;height=55.4
</details>

In [17]:
# För den dyraste bilen, ta fram width, height, length:
# använder loc + index för den dyraste bilen, som jag får fram med idxmax() metoden.
width = autos.loc[autos['price'].idxmax(),'width']
height = autos.loc[autos['price'].idxmax(), 'height']
length = autos.loc[autos['price'].idxmax(), 'length']
print(f'Length is {length}, height is {height}, width is {width}, of the most expensive car.')

Length is 199.2, height is 55.4, width is 72.0, of the most expensive car.


In [48]:
# ett annat sätt att lösa uppgiften:
the_most_expensive_car = autos.loc[autos['price'] == autos['price'].max()] # returnerar hela raden
the_most_expensive_car
# my_df = the_most_expensive_car[['width','height','length']]
# print(my_df)
# width = my_df['width']
# width

aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
74        std    hardtop   3.8        14                8.0         3715   

   drive-wheels engine-location  engine-size engine-type  ...  \
74          rwd           front          304        ohcv  ...   

   normalized-losses num-of-cylinders  num-of-doors  peak-rpm    price  \
74               NaN            eight           two    4500.0  45400.0   

    stroke symboling  wheel-base width price-per-hp  
74    3.35         1       112.0  72.0        246.7  

[1 rows x 27 columns]

**Q4:** What is the lowest price per horsepower in the dataset, and what brand ("make") is that car?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;price per horsepower=72.84
&nbsp;&nbsp;&nbsp;brand=Toyota
</details>

In [41]:
autos['price-per-hp'] = round(autos['price'] / autos['horsepower'],1)
lowest_price_per_hp = autos['price-per-hp'].min()
print(f'lägsta pris per hk: {lowest_price_per_hp}.')

autos.loc[autos['price-per-hp'] == autos['price-per-hp'].min()]['make']

lägsta pris per hk: 72.8.


167    toyota
Name: make, dtype: object

**Q5:** How many of the cars in the dataset has as many cylinders as they have doors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;cars=95
</details>

In [49]:
autos[(autos['num-of-cylinders'] == autos['num-of-doors']]


aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
3          std      sedan  3.19        24               10.0         2337   
11         std      sedan  3.50        23                8.8         2395   
20         std      sedan  3.03        38                9.6         1909   
24         std  hatchback  2.97        31                9.4         1967   
25         std      sedan  2.97        31                9.4         1989   
..         ...        ...   ...       ...                ...          ...   
198      turbo      sedan  3.62        17                7.5         3045   
199      turbo      wagon  3.62        17                7.5         3157   
200        std      sedan  3.78        23                9.5         2952   
201      turbo      sedan  3.78        19                8.7         3049   
204      turbo      sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type  ...  \
3            fwd           front          109         ohc  ...   
11           rwd           front          108         ohc  ...   
20           fwd           front           90         ohc  ...   
24           fwd           front           90         ohc  ...   
25           fwd           front           90         ohc  ...   
..           ...             ...          ...         ...  ...   
198          rwd           front          130         ohc  ...   
199          rwd           front          130         ohc  ...   
200          rwd           front          141         ohc  ...   
201          rwd           front          141         ohc  ...   
204          rwd           front          141         ohc  ...   

    normalized-losses num-of-cylinders  num-of-doors  peak-rpm    price  \
3               164.0             four          four    5500.0  13950.0   
11              192.0             four          four    5800.0  16925.0   
20               81.0             four          four    5400.0   6575.0   
24              148.0             four          four    5500.0   6229.0   
25              148.0             four          four    5500.0   6692.0   
..                ...              ...           ...       ...      ...   
198             103.0             four          four    5100.0  18420.0   
199              74.0             four          four    5100.0  18950.0   
200              95.0             four          four    5400.0  16845.0   
201              95.0             four          four    5300.0  19045.0   
204              95.0             four          four    5400.0  22625.0   

     stroke symboling  wheel-base width price-per-hp  
3      3.40         2        99.8  66.2        136.8  
11     2.80         0       101.2  64.8        167.6  
20     3.11         0        94.5  63.6         93.9  
24     3.23         1        93.7  63.8         91.6  
25     3.23         1        93.7  63.8         98.4  
..      ...       ...         ...   ...          ...  
198    3.15        -2       104.3  67.2        113.7  
199    3.15        -1       104.3  67.2        117.0  
200    3.15        -1       109.1  68.9        147.8  
201    3.15        -1       109.1  68.8        119.0  
204    3.15        -1       109.1  68.9        198.5  

[95 rows x 27 columns]

**Use the "autos" dataset and write python code to solve the following tasks:**

**T1:** Print a string to inform the user of the price difference between the cheapest and the most expensive car in the dataset.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;The price difference between the cheapest and the most expensive car is 40282.0
</details>

In [46]:
cheapest = autos['price'].min()
most_expensive = autos['price'].max()

print(f'Price delta between cheapest and most expensive car is {most_expensive - cheapest} $')

Price delta between cheapest and most expensive car is 40282.0 $


**T2:** Ask the user to input a brand, then print the price range for that brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volvo<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'volvo' ranges from 12940.0 to 22625.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: toyota<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'toyota' ranges from 5348.0 to 17669.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 3:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [ ]:
brand = input('Input brand / make: ')


**T3:** Ask the user to input a brand, then print the number of cars in the dataset for that brand, and all attributes for a random sample car of that brand.
<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: mazda<br>
&nbsp;&nbsp;&nbsp;There are 17 cars of brand 'mazda' in the dataset.<br><br>
&nbsp;&nbsp;&nbsp;Here is the data for a random 'mazda' car:<br>
&nbsp;&nbsp;&nbsp;aspiration = std<br>
&nbsp;&nbsp;&nbsp;body-style = sedan<br>
&nbsp;&nbsp;&nbsp;bore = 3.03<br>
&nbsp;&nbsp;&nbsp;city-mpg = 31<br>
&nbsp;&nbsp;&nbsp;compression-ratio = 9.0<br>
&nbsp;&nbsp;&nbsp;curb-weight = 1945<br>
&nbsp;&nbsp;&nbsp;drive-wheels = fwd<br>
&nbsp;&nbsp;&nbsp;engine-location = front<br>
&nbsp;&nbsp;&nbsp;engine-size = 91<br>
&nbsp;&nbsp;&nbsp;engine-type = ohc<br>
&nbsp;&nbsp;&nbsp;fuel-system = 2bbl<br>
&nbsp;&nbsp;&nbsp;fuel-type = gas<br>
&nbsp;&nbsp;&nbsp;height = 54.1<br>
&nbsp;&nbsp;&nbsp;highway-mpg = 38<br>
&nbsp;&nbsp;&nbsp;horsepower = 68.0<br>
&nbsp;&nbsp;&nbsp;length = 166.8<br>
&nbsp;&nbsp;&nbsp;make = mazda<br>
&nbsp;&nbsp;&nbsp;normalized-losses = 113.0<br>
&nbsp;&nbsp;&nbsp;num-of-cylinders = four<br>
&nbsp;&nbsp;&nbsp;num-of-doors = four<br>
&nbsp;&nbsp;&nbsp;peak-rpm = 5000.0<br>
&nbsp;&nbsp;&nbsp;price = 6695.0<br>
&nbsp;&nbsp;&nbsp;stroke = 3.15<br>
&nbsp;&nbsp;&nbsp;symboling = 1<br>
&nbsp;&nbsp;&nbsp;wheel-base = 93.1<br>
&nbsp;&nbsp;&nbsp;width = 64.2<br>
</details>

**T4:** Ask the user to input a brand, then export all cars of that brand to a .csv file with the same name as the brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volkswagen<br>
&nbsp;&nbsp;&nbsp;Exported 12 cars to 'volkswagen.csv'<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>